### ISSUE TITLE
MueLu: Build errors for Scalar=std::complex<double>

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#1339]
@trilinos/muelu 

The build errors are in `muelu/src/Graph/MueLu_VariableDofLaplacianFactory_def.hpp` (and possibly elsewhere).  They relate to conversions from Scalar to built-in integer types like `int`.  This is not defined when Scalar is complex.  Best practice is to take the real part via `Teuchos::ScalarTraits<Scalar>::real(someScalar)`, then assign to the integer type.  There are also comparisons of Scalar with `int` that need fixing (presumably in the same way).

You folks aren't assuming Scalar is always 64 bits, are you?  That seems unwise....

**Comment by jhux2** [comment]
Are you using an up-to-date develop?  @tawiesn pushed at least one fix with commit 67c28f94c.

**Comment by etphipp** [comment]
That approach won't work for stokhos either.  I think people have used Teuchos::as for these types of conversions.

-Eric

On May 17, 2017, at 5:42 PM, Mark Hoemmen <notifications@github.com<mailto:notifications@github.com>> wrote:


@trilinos/muelu<https://github.com/orgs/trilinos/teams/muelu>

The build errors are in muelu/src/Graph/MueLu_VariableDofLaplacianFactory_def.hpp (and possibly elsewhere). They relate to conversions from Scalar to built-in integer types like int. This is not defined when Scalar is complex. Best practice is to take the real part via Teuchos::ScalarTraits<Scalar>::real(someScalar), then assign to the integer type. There are also comparisons of Scalar with int that need fixing (presumably in the same way).

You folks aren't assuming Scalar is always 64 bits, are you? That seems unwise....

-

**Comment by mhoemmen** [comment]
Let me try again.  I was working on #941 and wanted to test MueLu.  Tried fixing one build error and saw there were a bunch.

**Comment by tawiesn** [comment]
working on it. There were more errors than i initially saw from the test output.

**Comment by tawiesn** [comment]
Checkin script is running...

**Comment by mhoemmen** [comment]
@tawiesn thanks for pushing the fix! :-D

**Comment by jhux2** [comment]
Can this issue be closed?

**Comment by tawiesn** [comment]
@jhux2 i referenced the issue in my commit. It will be automatically closed once the patch is moved to master.

**Comment by jhux2** [comment]
> @jhux2 i referenced the issue in my commit. It will be automatically closed once the patch is moved to master.

Ok, thanks!
